lib/google_contacts_api/api.rb

```diff
@@ -29,26 +29,101 @@ def get(link, params = {}, headers = {})
         raise UnauthorizedError if defined?(e.response) && self.class.parse_response_code(e.response) == 401
         raise e
       end
-      
+
       # OAuth 1.0 uses Net::HTTP internally
       raise UnauthorizedError if result.is_a?(Net::HTTPUnauthorized)
       result
     end
 
+    # these two methods will replace my previous two, but they aren't wired up to
+    # the rest of the code yet
     # Post request to specified link, with query params
-    # Not tried yet, might be issues with params
-    def post(link, params = {}, headers = {})
-      raise NotImplementedError
-      params["alt"] = "json"
-      @oauth.post("#{BASE_URL}#{link}?#{params.to_query}", headers)
+    def post_v2(link, params = {}, headers = {})
+      merged_params = params_with_defaults(params)
+      uri = "#{BASE_URL}#{link}?#{merged_params.to_query}"
+      begin
+        response = @oauth.post(uri, headers)
+      rescue => e
+        raise UnauthorizedError if defined?(e.response) && self.class.parse_response_code(e.response) == 401
+        raise e
+      end
     end
 
     # Put request to specified link, with query params
-    # Not tried yet
-    def put(link, params = {}, headers = {})
-      raise NotImplementedError
-      params["alt"] = "json"
-      @oauth.put("#{BASE_URL}#{link}?#{params.to_query}", headers)
+    def put_v2(link, params = {}, headers = {})
+      # Doesn't handle id yet
+      merged_params = params_with_defaults(params)
+      uri = "#{link}?#{merged_params.to_query}"
+      begin
+        response = @oauth.put(uri, headers)
+      rescue => e
+        raise UnauthorizedError if defined?(e.response) && self.class.parse_response_code(e.response) == 401
+        raise e
+      end
+    end
+
+    def post(contact)
+      options ={
+        headers: {
+          'Content-type' => 'application/json'
+          },
+        body: contact.entry_json
+      }
+      merged_params = params_with_defaults(options)
+      uri = URI.parse("https://www.google.com/m8/feeds/contacts/default/full/")
+
+      response = @oauth.post(uri, merged_params)
+    end
+
+    # Put request to specified link, with query params
+    def put(contact)
+      options ={
+        headers: {
+          'Content-type' => 'application/json',
+          'If-Match' => '*'
```

Comment:
This header ignores version history tags.  The point of those tags is stop overwriting any changes that have happened to the contact since we last accessed it.